### PR TITLE
add time selection after which files should be deleted

### PR DESCRIPTION
**Pull request by sualko**
This change should be backward compatible.

This comes handy if you want to delete files e.g. X days after the last modification. It's probably still confusing for the user, because upload time (creation time) is only available if the file was uploaded via webdav, but I don't have enough information about the file handling to fix this issue.

We need this enhancement for the outlook addon from Sendent.

Todo:
- [x] adapt tests

cc @usselite

### CODE REVIEW

File: lib/Migration/Version011000Date20210206170027.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		if ($schema->hasTable('retention')) {
+			$table = $schema->getTable('retention');
+
+			$table->addColumn('time_after', 'integer', [
+				'notnull' => true,
```
Comment (rullzer) [inline lib/Migration/Version011000Date20210206170027.php:31]: @nickvergessen does this work with Oracle?

Comment (nickvergessen) [reply lib/Migration/Version011000Date20210206170027.php:31]: No because the default is 0 and 0 on int is "null"

File: templates/admin.php [hunk 2]
```diff
@@ -73,6 +74,13 @@
 		<option value="3"><?php p($l->t('Years')); ?></option>
 	</select>
 
+	<?php p($l->t('after')); ?>
+
+	<select id="retention_after">
+		<option value="0"><?php p($l->t('creation')); ?></option>
+		<option value="1"><?php p($l->t('last modification')); ?></option>
```
Comment (nickvergessen) [inline templates/admin.php:81]:
```suggestion
		<option value="0"><?php p($l->t('Creation')); ?></option>
		<option value="1"><?php p($l->t('Last modification')); ?></option>
```

File: lib/Migration/Version011000Date20210206170027.php (hunk 1 above)
Comment (rullzer) [reply lib/Migration/Version011000Date20210206170027.php:31]: As I thought.

@sualko can you make it nullable. And add the logic that if it is null you just assume 0?
(else the oracle DB fails hard).

Comment (sualko) [reply lib/Migration/Version011000Date20210206170027.php:31]: Since I use a `(int)` cast everywhere, that should be no issue.

Comment (rullzer) [reply lib/Migration/Version011000Date20210206170027.php:31]: But then it should also be nullable. Else oracle complains (0 === null in oracles twisted world)

Comment (nickvergessen) [reply lib/Migration/Version011000Date20210206170027.php:31]: yeah thats what I understood, when default is 0 you need to be nullable

Comment (sualko) [reply lib/Migration/Version011000Date20210206170027.php:31]: Seams to be fun to be working with Oracle. I will change that asap.

Comment (rullzer) [reply lib/Migration/Version011000Date20210206170027.php]: You have no idea :P

**Review by rullzer — APPROVED** [review]

**Comment by rullzer** [comment]
Ah shit conflicts.
Could you rebase upon master?

**Comment by sualko** [comment]
> Could you rebase upon master?

Done

**Comment by max-nextcloud** [comment]
@sualko I am revisiting this because currently there's no way to delete files a given amount of time after the upload OR changes to the file uploaded file were made. If one selects `MTIME` the `MTIME` before the upload will also count. If one selects `CTIME` changes after the upload won't count.

Did you need both settings for the outlook addon?

Which behavior do you rely upon?

I'm considering to remove the setting and simplify this to n days after `MAX(MTIME, CTIME, uploadtime)`.

Alternatively we might take the upload time into account also for MTIME. I.e. count the upload as a modification.

**Comment by usselite** [comment]
Hi @max-nextcloud I think removing this is OK, I am not familiar with the Nextcloud app code itself but this is one of the components where it was built for: https://github.com/Sendent-B-V/Sendent-App-for-Nextcloud/blob/master/lib/Listener/ShareCreatedListener.php

Basically what flow does, when share expires or when share is deleted, automatically clean-up after X days. 

If I understand correctly, your modification won't be an issue, as if a existing share is modified, neither flows will be triggered until it actually is expired or deleted. Is that correct?

**Comment by max-nextcloud** [comment]
> If I understand correctly, your modification won't be an issue, as if a existing share is modified, neither flows will be triggered until it actually is expired or deleted. Is that correct?

Looking at the Sendent app it uses tags to indicate the expiry / mark files for deletion. This mechanism would not change at all. It's just the time calculation for that files that are tagged that would be affected.

Sendent seems to make use of the timestamps here in particular:

https://github.com/Sendent-B-V/Sendent-App-for-Nextcloud/blob/0dc8a8fdb63f36703612edb6464675abf45ca7b8/lib/Listener/ShareDeletedListener.php#L120-L121

So the file is touched - i.e. MTIME is updated in order to calculate the expiry correctly. This would not work if the upload time would be used instead.

With my proposed change the `MAX(MTIME, CTIME, uploadTime)` would be used. So the touch would still work as it does right now.

**Pull request merged**